komposable-architecture/src/main/java/com/toggl/komposable/extensions/EffectCancellationExtensions.kt

```diff
@@ -0,0 +1,51 @@
+package com.toggl.komposable.extensions
+
+import com.toggl.komposable.architecture.Effect
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.currentCoroutineContext
+import kotlinx.coroutines.flow.cancellable
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.onStart
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+private val mutex = Mutex()
+private val cancellationJobs: MutableMap<Any, MutableSet<Job>> = mutableMapOf()
+
+/**
+ * Makes an Effect cancellable by associating it with a given ID.
+ * If 'cancelInFlight' is true, any previous effects with the same ID will be cancelled.
+ *
+ * @param id The identifier for this cancellable effect. Used to manage cancellation.
+ * @param cancelInFlight If true, cancels any ongoing effects with the same ID.
+ * @return A new Effect instance that is cancellable.
+ */
+fun <Action> Effect<Action>.cancellable(id: Any, cancelInFlight: Boolean = false): Effect<Action> =
+    Effect {
+        this().onStart {
+            mutex.withLock {
```

Comment:
Isn't all the flows from `Effect`s are collected in the same thread (the main dispatcher)? If so, do we need the mutex?

Reply:
Tbh, I'm not an expert here, this was inspired by https://github.com/wearemakery/kotlin-composable-architecture/blob/master/composable-architecture/src/main/kotlin/composablearchitecture/Effect%2BCancellation.kt I used mutex as well just to be sure :-D But my understanding is that effects can indeed run in parallel thus mutex might be necessary. I guess we should try to test this 🤔